In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import sqlite3
from sqlite3 import Error


# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

In [2]:
from IPython.display import display
pd.set_option('display.max_columns', None)

In [3]:
data_path = "factbook.db"

In [4]:
connection = sqlite3.connect(data_path)

# 1. What is DBMS
DBMS: Database Management System is a software system that uses a standard method of cataloging, retrieving, and running queries on data. Some DBMS examples include MySQL, PostgreSQL, Microsoft Access, SQL Server, FileMaker, Oracle, RDBMS, dBASE, Clipper, and FoxPro

A tpical relatable example is wtfellows.db that stores datasets of factsbooks.

# 2. What is RDBMS
RDBMS: Relational Database Management System organizes data into tables which can be linked—or related—based on data common to each. This capability enables you to retrieve an entirely new table from data in one or more tables with a single query. 

For example, imagine your company maintains a customer table that contains company data about each customer account and one or more transaction tables that contain data describing individual transactions.

# 3. What is SQL
SQL: You can communicate with relational databases using Structured Query Language (SQL), the standard language for interacting with management systems. SQL allows the joining of tables using a few lines of code, with a structure most nontechnical employees can learn quickly. Example of a sql is :

In [5]:
tables = pd.read_sql("""

SELECT *
FROM sqlite_master
WHERE type = 'table';

""", connection)

tables

type             name         tbl_name  rootpage  \
0  table  sqlite_sequence  sqlite_sequence         3   
1  table            facts            facts        47   
2  table           cities           cities         2   

                                                 sql  
0             CREATE TABLE sqlite_sequence(name,seq)  
1  CREATE TABLE "facts" ("id" INTEGER PRIMARY KEY...  
2  CREATE TABLE cities (\n        id integer prim...

Here we can see that the database *factbook.db* has two tables. Namely:
1. Facts
2. Cities

# 4. What is a Database
A database is a systematic collection of data, that supports electronic storage and manipulation of data. Example is given below:

In [6]:
pd.read_sql("""

SELECT *
FROM facts;

""", connection)

id code            name       area  area_land  area_water    population  \
0      1   af     Afghanistan   652230.0   652230.0         0.0  3.256434e+07   
1      2   al         Albania    28748.0    27398.0      1350.0  3.029278e+06   
2      3   ag         Algeria  2381741.0  2381741.0         0.0  3.954217e+07   
3      4   an         Andorra      468.0      468.0         0.0  8.558000e+04   
4      5   ao          Angola  1246700.0  1246700.0         0.0  1.962535e+07   
..   ...  ...             ...        ...        ...         ...           ...   
256  257   zh  Atlantic Ocean        NaN        NaN         NaN           NaN   
257  258   xo    Indian Ocean        NaN        NaN         NaN           NaN   
258  259   zn   Pacific Ocean        NaN        NaN         NaN           NaN   
259  260   oo  Southern Ocean        NaN        NaN         NaN           NaN   
260  261   xx           World        NaN        NaN         NaN  7.256490e+09   

     population_growth  birth_rate  death_rate  migration_rate  
0                 2.32       38.57       13.89            1.51  
1                 0.30       12.92        6.58            3.30  
2                 1.84       23.67        4.31            0.92  
3                 0.12        8.13        6.96            0.00  
4                 2.78       38.78       11.49            0.46  
..                 ...         ...         ...             ...  
256                NaN         NaN         NaN             NaN  
257                NaN         NaN         NaN             NaN  
258                NaN         NaN         NaN             NaN  
259                NaN         NaN         NaN             NaN  
260               1.08       18.60        7.80             NaN  

[261 rows x 11 columns]

# 5. What are tables and fields
- The data in RDBMS is stored in database objects called tables. A table is a collection of related data entries and it consists of columns and rows.
- Every table is broken up into smaller entities called fields. The fields in the facts table consist of id, code, name, area, area_land, area_water, population, population_growth, birth_rate, death_rate, and migration_rate. 
- A field is a column in a table that is designed to maintain specific information about every record in the table. The table has 11 fields

- A record, also called a row, is each individual entry that exists in a table. For example, there are 261 records in the above facts table. A record is a horizontal entity in a table.

- A column is a vertical entity in a table that contains all information associated with a specific field in a table.

Now lets us view the "*cities*" table

In [7]:
pd.read_sql("""

SELECT *
FROM cities;

""", connection)

id          name  population  capital facts_id
0      1    Oranjestad       37000        1      216
1      2  Saint John'S       27000        1        6
2      3     Abu Dhabi      942000        1      184
3      4         Dubai     1978000        0      184
4      5       Sharjah      983000        0      184
..   ...           ...         ...      ...      ...
392  393       Mbabane       66000        1      167
393  394         Sanaa     2419000        1      193
394  395          Aden      784000        0      193
395  396        Lusaka     1802000        1      194
396  397        Harare     1542000        1      195

[397 rows x 5 columns]

### Above we can tell that:
1. there are 11 fields for *facts* tables and 
2. there are 5 fields for *cities* table

# 6. What is a Primary Key
A primary key is a column or a set of columns in a table whose values uniquely identify a row in the table. 

In [8]:
tables = pd.read_sql("""

SELECT *
FROM sqlite_master
WHERE type = 'table';

""", connection)

tables['sql']

0               CREATE TABLE sqlite_sequence(name,seq)
1    CREATE TABLE "facts" ("id" INTEGER PRIMARY KEY...
2    CREATE TABLE cities (\n        id integer prim...
Name: sql, dtype: object

### From the above we can see that for both tables "facts" and "cities", the primary key is "id". And there are no two record with the same id.

# 7. What is a unique key?
A unique key is a set of one or more than one fields/columns of a table that uniquely identify a record in a database table. One can say that it is little like primary key but it can accept only one null value and it cannot have duplicate values.

Here we set the email Address as the unique key:

In [9]:
## Creating a database to store our tables
## Defining a function that helps us communicate without our database without having to call everytime
data_path2 = 'womentechsters.db'
conn = sqlite3.connect(data_path2)
def execute_sql(conn, query):
    cursor = conn.cursor()
    try:
        cursor.execute(query)
        conn.commit()
        print("SQL Executed Successfully")
    except Error as e:
        print (f"The error '{e}' occurred")

In [10]:
create_table = """
CREATE TABLE Fellows(
    id integer PRIMARY KEY, 
    fName text, 
    lName text, 
    learningTrack text, 
    employmentStat text, 
    instructorName text, 
    emailAdd text not null unique, 
    orientationAttend text
    );

"""
execute_sql(conn, create_table)

The error 'table Fellows already exists' occurred


In [11]:
def insertIntoTable(id,fName,lName,learningTrack,employmentStat,instructorName,emailAdd,orientationAttend):
    try:
        conn = sqlite3.connect('womentechsters.db')
        cursor = conn.cursor()
        print("Connected to SQLite")

        sqlite_insert = """INSERT INTO Fellows
                          (id,fName,lName,learningTrack,employmentStat,instructorName,emailAdd,orientationAttend) 
                          VALUES (?,?,?,?,?,?,?,?);"""

        data_tuple = (id,fName,lName,learningTrack,employmentStat,instructorName,emailAdd,orientationAttend)
        cursor.execute(sqlite_insert, data_tuple)
        conn.commit()
        print("Python Variables inserted successfully into Fellows table")

        cursor.close()

    except sqlite3.Error as error:
        print("Failed to insert Python variable into sqlite table", error)
    finally:
        if conn:
            conn.close()
            print("The SQLite connection is closed")


insertIntoTable(1,'Anita','Mburu','Data Science', 'Employed', 'Winner Emeto', 'anita.mburu@wtf.org', 'Yes')
insertIntoTable(2,'Chinenye','James','Data Science', 'Self Employed', 'Winner Emeto', 'chi.james@wtf.org', 'Yes')
insertIntoTable(3,'Rukayat','Fagbola','Software Engineering','Self Employed','Naomi Praise', 'rukky.fags@wtf.org', 'No')
insertIntoTable(4,'Chioma','Obi','Cybersecurity', 'Unemployed','Adams Onoms', 'chi.obi@wtf.org', 'Yes')
insertIntoTable(5,'Angelica','Dairo','Product Management', 'Employed','Salima Wartz', 'ang.dairo@wtf.org', 'Yes')
insertIntoTable(6,'Joy','Kareko','Data Science', 'Self Employed','Winner Emeto', 'joy.kare@wtf.org', 'No')
insertIntoTable(7,'Choice','Iruh','Data Science', 'Employed', 'Winner Emeto','choic.iruh@wtf.org', 'No')

Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: Fellows.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: Fellows.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: Fellows.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: Fellows.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: Fellows.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: Fellows.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: Fellows.id
The SQLite connection is closed


In [12]:
fellows = pd.read_sql("""SELECT *
                        FROM Fellows;""", conn)
fellows

id     fName    lName         learningTrack employmentStat instructorName  \
0   1     Anita    Mburu          Data Science       Employed   Winner Emeto   
1   2  Chinenye    James          Data Science  Self Employed   Winner Emeto   
2   3   Rukayat  Fagbola  Software Engineering  Self Employed   Naomi Praise   
3   4    Chioma      Obi         Cybersecurity     Unemployed    Adams Onoms   
4   5  Angelica    Dairo    Product Management       Employed   Salima Wartz   
5   6       Joy   Kareko          Data Science  Self Employed   Winner Emeto   
6   7    Choice     Iruh          Data Science       Employed   Winner Emeto   

              emailAdd orientationAttend  
0  anita.mburu@wtf.org               Yes  
1    chi.james@wtf.org               Yes  
2   rukky.fags@wtf.org                No  
3      chi.obi@wtf.org               Yes  
4    ang.dairo@wtf.org               Yes  
5     joy.kare@wtf.org                No  
6   choic.iruh@wtf.org                No

# 8. What is a foreign key
A foreign key is a column or group of columns in a relational database table that provides a link between data in two tables. It acts as a cross-reference between tables because it references the primary key of another table, thereby establishing a link between them.

Below we will see that the foreign key in table datascience is the FellowID. 

In [13]:
create_table2 = """
CREATE TABLE studentrack(
    id text PRIMARY KEY, 
    fName text, 
    lName text, 
    learningTrack text
    );
"""

In [14]:
execute_sql(conn, create_table2)

The error 'table studentrack already exists' occurred


In [15]:
def insertIntoTable2(id,fName,lName,learningTrack):
    try:
        conn = sqlite3.connect('womentechsters.db')
        cursor = conn.cursor()
        print("Connected to SQLite")

        sqlite_insert = """INSERT INTO studentrack(id,fName,lName,learningTrack)
                          VALUES (?,?,?,?);"""

        data_tuple = (id,fName,lName,learningTrack)
        cursor.execute(sqlite_insert, data_tuple)
        conn.commit()
        print("Python Variables inserted successfully into studentrack table")

        cursor.close()

    except sqlite3.Error as error:
        print("Failed to insert Python variable into sqlite table", error)
    finally:
        if conn:
            conn.close()
            print("The SQLite connection is closed")


insertIntoTable2('WT/21/022','Anita','Mburu','Data Science')
insertIntoTable2('WT/21/049','Chinenye','James','Data Science'),
insertIntoTable2('WT/21/001','Rukayat','Fagbola','Software Engineering'),
insertIntoTable2('WT/21/056','Chioma','Obi','Cybersecurity'),
insertIntoTable2('WT/21/010','Angelica','Dairo','Product Management'),
insertIntoTable2('WT/21/118','Joy','Kareko','Data Science'),
insertIntoTable2('WT/21/051','Choice','Iruh','Data Science')

Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: studentrack.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: studentrack.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: studentrack.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: studentrack.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: studentrack.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: studentrack.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: studentrack.id
The SQLite connection is

In [16]:
stdtrck = pd.read_sql("""SELECT *
                        FROM studentrack;""", conn)
stdtrck

id     fName    lName         learningTrack
0  WT/21/022     Anita    Mburu          Data Science
1  WT/21/049  Chinenye    James          Data Science
2  WT/21/001   Rukayat  Fagbola  Software Engineering
3  WT/21/056    Chioma      Obi         Cybersecurity
4  WT/21/010  Angelica    Dairo    Product Management
5  WT/21/118       Joy   Kareko          Data Science
6  WT/21/051    Choice     Iruh          Data Science

In [17]:
## Create another table within the database womentecster.db
create_table3 = """
CREATE TABLE track(
    id text PRIMARY KEY, 
    learningTrack text,  
    FOREIGN KEY(id) REFERENCES track_fellows(id)
    );
    
"""


In [18]:
execute_sql(conn, create_table3)

The error 'table track already exists' occurred


In [19]:
def insertIntoTable3(id,learningTrack):
    try:
        conn = sqlite3.connect('womentechsters.db')
        cursor = conn.cursor()
        print("Connected to SQLite")

        sqlite_insert = """INSERT INTO track(id,learningTrack)
                          VALUES (?,?);"""

        data_tuple = (id,learningTrack)
        cursor.execute(sqlite_insert, data_tuple)
        conn.commit()
        print("Python Variables inserted successfully into studentrack table")

        cursor.close()

    except sqlite3.Error as error:
        print("Failed to insert Python variable into sqlite table", error)
    finally:
        if conn:
            conn.close()
            print("The SQLite connection is closed")


insertIntoTable3('WT/21/022', 'Data Science')
insertIntoTable3('WT/21/049', 'Data Science')
insertIntoTable3('WT/21/001', 'Software Engineering')
insertIntoTable3('WT/21/056', 'Cybersecurity')
insertIntoTable3('WT/21/010', 'Product Management')
insertIntoTable3('WT/21/118', 'Data Science')
insertIntoTable3('WT/21/051', 'Data Science')

Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: track.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: track.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: track.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: track.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: track.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: track.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: track.id
The SQLite connection is closed


In [20]:
trck = pd.read_sql("""SELECT *
                        FROM track;""", conn)
trck

id         learningTrack
0  WT/21/022          Data Science
1  WT/21/049          Data Science
2  WT/21/001  Software Engineering
3  WT/21/056         Cybersecurity
4  WT/21/010    Product Management
5  WT/21/118          Data Science
6  WT/21/051          Data Science

# 9. What is Join?
JOIN is used when you want to connect two tables to each other. It works when you have a common key in each of them. Understanding the concept of Keys is crucial for connecting (joining) between data set (tables). A key is uniquely identifies each record (row) in a table. It can consinst of one value (cell) - usually ID, or from a combination of values that are unique in the table.

When joinin between different tables, you must:

- Decide what type of join to use.

In [21]:
join_tables = """
SELECT * 
FROM facts
JOIN cities 
ON cities.id = facts.id;

"""
join_result = pd.read_sql(join_tables, connection)
join_result

id code            name       area  area_land  area_water    population  \
0      1   af     Afghanistan   652230.0   652230.0         0.0  3.256434e+07   
1      2   al         Albania    28748.0    27398.0      1350.0  3.029278e+06   
2      3   ag         Algeria  2381741.0  2381741.0         0.0  3.954217e+07   
3      4   an         Andorra      468.0      468.0         0.0  8.558000e+04   
4      5   ao          Angola  1246700.0  1246700.0         0.0  1.962535e+07   
..   ...  ...             ...        ...        ...         ...           ...   
256  257   zh  Atlantic Ocean        NaN        NaN         NaN           NaN   
257  258   xo    Indian Ocean        NaN        NaN         NaN           NaN   
258  259   zn   Pacific Ocean        NaN        NaN         NaN           NaN   
259  260   oo  Southern Ocean        NaN        NaN         NaN           NaN   
260  261   xx           World        NaN        NaN         NaN  7.256490e+09   

     population_growth  birth_rate  death_rate  migration_rate   id  \
0                 2.32       38.57       13.89            1.51    1   
1                 0.30       12.92        6.58            3.30    2   
2                 1.84       23.67        4.31            0.92    3   
3                 0.12        8.13        6.96            0.00    4   
4                 2.78       38.78       11.49            0.46    5   
..                 ...         ...         ...             ...  ...   
256                NaN         NaN         NaN             NaN  257   
257                NaN         NaN         NaN             NaN  258   
258                NaN         NaN         NaN             NaN  259   
259                NaN         NaN         NaN             NaN  260   
260               1.08       18.60        7.80             NaN  261   

              name  population  capital facts_id  
0       Oranjestad       37000        1      216  
1     Saint John'S       27000        1        6  
2        Abu Dhabi      942000        1      184  
3            Dubai     1978000        0      184  
4          Sharjah      983000        0      184  
..             ...         ...      ...      ...  
256          Abuja     2153000        1      129  
257  Port Harcourt     1894000        0      129  
258         Kaduna     1524000        0      129  
259      Amsterdam     1056000        1      125  
260      Rotterdam     1014000        0      125  

[261 rows x 16 columns]

# 10. What are the types of join and explain each.
 The most common are:
- (INNER) JOIN - keep only records that match the condition (after the ON) in both the tables, and records in both tables that do not match wouldn't appear in the output
- LEFT (OUTER) JOIN - keep all the values from the first (left) table - in conjunction with the matching rows from the right table. The columns from the right table, that don't have matching value in the left, would have NULL values.
- RIGHT (OUTER) JOIN - this returns all records from the right table, and matched records from the left table
- FULL (OUTER) JOIN - returns all records/rows when there is a match in either left or right table.
- CROSS JOIN - this produces a table that merges each row from the first table with each second table row. It is also not required to include the CROSS JOIN query when using this query language.
- SELF JOIN - This creates a table by joining itself and temporarily names at least one table in an SQL Statement

In [22]:
students = """
CREATE TABLE Student(id integer PRIMARY KEY, name text, age integer, class_s text);
   """

In [23]:
#execute_sql(conn, students_table)
execute_sql(conn, students)

The error 'table Student already exists' occurred


In [24]:
def insertIntoTable4(id,name,age,class_s):
    try:
        conn = sqlite3.connect('womentechsters.db')
        cursor = conn.cursor()
        print("Connected to SQLite")

        sqlite_insert = """INSERT INTO Student(id,name,age,class_s)
                          VALUES (?,?,?,?);"""

        data_tuple = (id,name,age,class_s)
        cursor.execute(sqlite_insert, data_tuple)
        conn.commit()
        print("Python Variables inserted successfully into student table")

        cursor.close()

    except sqlite3.Error as error:
        print("Failed to insert Python variable into sqlite table", error)
    finally:
        if conn:
            conn.close()
            print("The SQLite connection is closed")


insertIntoTable4(1,'Amara Iloh',10, 'SS1')
insertIntoTable4(2,'Chinenye James', 11, 'SS2')
insertIntoTable4(3,'Rukayat Fagbola', 14, 'SS3')
insertIntoTable4(4,'Chioma Obi', 12, 'SS1')


Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: Student.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: Student.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: Student.id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: Student.id
The SQLite connection is closed


In [25]:
stds = pd.read_sql("""SELECT *
                        FROM Student;""", conn)
stds

id             name  age class_s
0   1       Amara Iloh   10     SS1
1   2   Chinenye James   11     SS2
2   3  Rukayat Fagbola   14     SS3
3   4       Chioma Obi   12     SS1

In [26]:
students_table2 = """
CREATE TABLE Studentdetails(
    reg_id integer PRIMARY KEY, 
    class_type text,
    serial_no interger,
    fees_amount integer
    );

"""

In [27]:
execute_sql(conn, students_table2)

The error 'table Studentdetails already exists' occurred


In [28]:
def insertIntoTable5(reg_id,class_type,serial_no,fees_amount):
    try:
        conn = sqlite3.connect('womentechsters.db')
        cursor = conn.cursor()
        print("Connected to SQLite")

        sqlite_insert = """INSERT INTO Studentdetails(reg_id,class_type,serial_no,fees_amount)
                          VALUES (?,?,?,?);"""

        data_tuple = (reg_id,class_type,serial_no,fees_amount)
        cursor.execute(sqlite_insert, data_tuple)
        conn.commit()
        print("Python Variables inserted successfully into studentdetails table")

        cursor.close()

    except sqlite3.Error as error:
        print("Failed to insert Python variable into sqlite table", error)
    finally:
        if conn:
            conn.close()
            print("The SQLite connection is closed")


insertIntoTable5(1,'Science', 1, 120000),
insertIntoTable5(2,'Commercial', 2, 130000),
insertIntoTable5(3,'Commercial', 2, 140000),
insertIntoTable5(4,'Art',3, 80000)


Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: Studentdetails.reg_id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: Studentdetails.reg_id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: Studentdetails.reg_id
The SQLite connection is closed
Connected to SQLite
Failed to insert Python variable into sqlite table UNIQUE constraint failed: Studentdetails.reg_id
The SQLite connection is closed


In [29]:
stdetails = pd.read_sql("""SELECT *
                        FROM Studentdetails;""", conn)
stdetails

reg_id  class_type  serial_no  fees_amount
0       1     Science          1       120000
1       2  Commercial          2       130000
2       3  Commercial          2       140000
3       4         Art          3        80000

### LEFT (OUTER) JOIN

In [30]:
left_join = """
SELECT id,name,fees_amount,class_type 
FROM Student
LEFT JOIN Studentdetails 
ON Student.id=Studentdetails.serial_no;

"""
left_join_result = pd.read_sql(left_join, conn)
left_join_result

id             name  fees_amount  class_type
0   1       Amara Iloh     120000.0     Science
1   2   Chinenye James     130000.0  Commercial
2   2   Chinenye James     140000.0  Commercial
3   3  Rukayat Fagbola      80000.0         Art
4   4       Chioma Obi          NaN        None

## RIGHT JOIN

In [31]:
right_join = """
SELECT id,name,fees_amount,class_type 
FROM Student
RIGHT JOIN Studentdetails 
ON Student.id=Studentdetails.serial_no;

"""
right_join_result = pd.read_sql(right_join, conn)
right_join_result

DatabaseError: Execution failed on sql '
SELECT id,name,fees_amount,class_type 
FROM Student
RIGHT JOIN Studentdetails 
ON Student.id=Studentdetails.serial_no;

': RIGHT and FULL OUTER JOINs are not currently supported

### FULL JOIN

In [32]:
full_join = """

SELECT * 
FROM Student
FULL OUTER JOIN Studentdetails 
ON Student.id=Studentdetails.serial_no;

"""
full_join_result = pd.read_sql(full_join, conn)
full_join_result

DatabaseError: Execution failed on sql '

SELECT * 
FROM Student
FULL OUTER JOIN Studentdetails 
ON Student.id=Studentdetails.serial_no;

': RIGHT and FULL OUTER JOINs are not currently supported

### 32. CROSS JOIN

In [33]:
cross_join = """
SELECT * 
FROM Student
CROSS JOIN Studentdetails;

"""
cross_join_result = pd.read_sql(cross_join, conn)
cross_join_result

id             name  age class_s  reg_id  class_type  serial_no  \
0    1       Amara Iloh   10     SS1       1     Science          1   
1    1       Amara Iloh   10     SS1       2  Commercial          2   
2    1       Amara Iloh   10     SS1       3  Commercial          2   
3    1       Amara Iloh   10     SS1       4         Art          3   
4    2   Chinenye James   11     SS2       1     Science          1   
5    2   Chinenye James   11     SS2       2  Commercial          2   
6    2   Chinenye James   11     SS2       3  Commercial          2   
7    2   Chinenye James   11     SS2       4         Art          3   
8    3  Rukayat Fagbola   14     SS3       1     Science          1   
9    3  Rukayat Fagbola   14     SS3       2  Commercial          2   
10   3  Rukayat Fagbola   14     SS3       3  Commercial          2   
11   3  Rukayat Fagbola   14     SS3       4         Art          3   
12   4       Chioma Obi   12     SS1       1     Science          1   
13   4       Chioma Obi   12     SS1       2  Commercial          2   
14   4       Chioma Obi   12     SS1       3  Commercial          2   
15   4       Chioma Obi   12     SS1       4         Art          3   

    fees_amount  
0        120000  
1        130000  
2        140000  
3         80000  
4        120000  
5        130000  
6        140000  
7         80000  
8        120000  
9        130000  
10       140000  
11        80000  
12       120000  
13       130000  
14       140000  
15        80000

### 31. SELF JOIN

In [34]:
self_join = """
SELECT facts.id,facts.name
FROM facts, cities
WHERE cities.population < facts.population;

"""
self_join_result = pd.read_sql(self_join, connection)
self_join_result

id         name
0        1  Afghanistan
1        1  Afghanistan
2        1  Afghanistan
3        1  Afghanistan
4        1  Afghanistan
...    ...          ...
62282  261        World
62283  261        World
62284  261        World
62285  261        World
62286  261        World

[62287 rows x 2 columns]

# 11. What is normalization?
Normalization involves the reducing of redundancy of data in a table to improve the integrity of the datasets. It is required to normalize our datasets because we are likely to faces issues such as:
- insertion anomaly: this occurs when we cannot insert data to a table without the presence of another attribute
- update anomaly: inconsistencies in data that result from data redundancy and a partial update of data
- deletion anomaly: happens when some attributes are lost because we deleted other attributes.

# 12. What is denormalization?
Denormalization is the process of adding redundant data to an otherwise normalized relational database to improve read performance of the database. It is also known as database optimization technique, where redundant data are added in the database to get rid of the complex join operations.

# 13A. What are the different normalization?
- 1NF: 1st Normal Form
- 2NF: 2nd Normal Form
- 3NF: 3rd Normal Form
- 4NF: 4th Normal Form
- 5NF: 5th Normal Form

NB:
- BCNF: Boyce-Codd Normal Form

# 13B. The normal forms are divided into five(5), explain:
### 1NF:
This prevents the problem of atomicity, which implies that values in a table should not be further divided. That is: a single cell should not hold multiple values. If a cell contains multiple values then the 1NF has been voilated.

### Second Normal Form (2NF):.
Meeting all requirements of the first normal form. Placing the subsets of data in separate tables and Creation of relationships between the tables using primary keys.

### Third Normal Form (3NF):.
This should meet all requirements of 2NF. Removing the columns which are not dependent on primary key constraints.

### Fourth Normal Form (4NF):.
Meeting all the requirements of third normal form and it should not have multi- valued dependencies.


# 14. What is a view?
A database view is a subset of a database and is based on a query that runs on one or more database tables. Database views are saved in the database as named queries and can be used to save frequently used, complex queries. There are two types of database views: 
- dynamic views, and 
- static views.

In [35]:
#Using the cities table from facts.db, we will try to view all capitals with 1.
view_example = """
CREATE VIEW [Positive Capital] AS
SELECT name, fees_amount
FROM Studentdetails
WHERE fees_amount < 100000;

"""
display_view = pd.read_sql(view_example, connection)
display_view

TypeError: 'NoneType' object is not iterable

# 15. What is an index?
An Index is a small table having only two columns. The first column comprises a copy of the primary or candidate key of a table. Its second column contains a set of pointers for holding the address of the disk block where that specific key value stored. Indexing is a data structure technique which allows you to quickly retrieve records from a database file.

In [36]:
create_index = """
CREATE INDEX checker 
ON cities

"""
view_tab = pd.read_sql(create_index, connection)
view_tab

DatabaseError: Execution failed on sql '
CREATE INDEX checker 
ON cities

': incomplete input

# 16A. What are the different types of index?
1. Single-column Indexes: are created based on only one table column
2. Unique Indexes: are used not only for performances but also for data integrity. A unique index does not allow any duplicates to be inserted into the table.
3. Composite Indexes: This is an index that is on two or more columns of the table
4. Implicit Indexes: these are indexes that are automatically created by the database server when an object is created for unique or primary key constraints.

# 16B. There are three types of indexes - List and explain
1. Clustering Index: In a clustered index, records themselves are stored in the Index and not pointers. Sometimes the Index is created on non-primary key columns which might not be unique for each record. In such a situation, you can group two or more columns to get the unique values and create an index which is called clustered Index. This also helps you to identify the record faster.
2. Primary Index: Primary Index is an ordered file which is fixed length size with two fields. The first field is the same a primary key and second, filed is pointed to that specific data block. In the primary Index, there is always one to one relationship between the entries in the index table.
3. Secondary Index: The secondary Index in DBMS can be generated by a field which has a unique value for each record, and it should be a candidate key. It is also known as a non-clustering index. This two-level database indexing technique is used to reduce the mapping size of the first level. For the first level, a large range of numbers is selected because of this; the mapping size always remains small.

# 17. What is a cursor?
A cursor can be viewed as a pointer to one row in a set of rows. The cursor can only reference one row at a time, but can move to other rows of the result set as needed.

To work with cursors you must use the following SQL statements:
- DECLARE CURSOR
- OPEN
- FETCH
- CLOSE
For instance: with the sqlite3 cursor we can execute, update rows in our database. Examples can be found above.

# 18. What is a relationship, and what are they?
A relationship is established between two database tables when one table uses a foreign key that references the primary key of another table. 
## For example: The studentrack and track table below, in the next line of code[](http://):

## Types of relationships:
- ONE-TO-ONE: This type of relationship allows only one record on each side of the relationship. The primary key relates to only one record (or none) in another table. For example, in a marriage, each spouse has only one other spouse. This kind of relationship can be implemented in a single table and therefore does not use a foreign key.

- ONE-TO-MANY: A one-to-many relationship allows a single record in one table to be related to multiple records in another table. Consider a business with a database that has Customers and Orders tables. A single customer can purchase multiple orders, but a single order could not be linked to multiple customers. Therefore the Orders table would contain a foreign key that matched the primary key of the Customers table, while the Customers table would have no foreign key pointing to the Orders table.

- MANY-TO-MANY: This is a complex relationship in which many records in a table can link to many records in another table. For example, our business probably needs Customers and Orders tables, and likely also needs a Products table.

In [37]:
trck

id         learningTrack
0  WT/21/022          Data Science
1  WT/21/049          Data Science
2  WT/21/001  Software Engineering
3  WT/21/056         Cybersecurity
4  WT/21/010    Product Management
5  WT/21/118          Data Science
6  WT/21/051          Data Science

In [38]:
stdtrck

id     fName    lName         learningTrack
0  WT/21/022     Anita    Mburu          Data Science
1  WT/21/049  Chinenye    James          Data Science
2  WT/21/001   Rukayat  Fagbola  Software Engineering
3  WT/21/056    Chioma      Obi         Cybersecurity
4  WT/21/010  Angelica    Dairo    Product Management
5  WT/21/118       Joy   Kareko          Data Science
6  WT/21/051    Choice     Iruh          Data Science

# 19. What is a query?
A database query is a request for data from a database. Usually the request is to retrieve data; however, data can also be manipulated using queries. The data can come from one or more tables, or even other queries. It is simply the language that SQL understands. Some queries include: 


In [39]:
insertIntoTable4(5,'Uju Okoli',16, 'SS3')

view_new = pd.read_sql("""SELECT *
                        FROM Student;""", conn)

view_new

Connected to SQLite
Python Variables inserted successfully into student table
The SQLite connection is closed


id             name  age class_s
0   1       Amara Iloh   10     SS1
1   2   Chinenye James   11     SS2
2   3  Rukayat Fagbola   14     SS3
3   4       Chioma Obi   12     SS1
4   5        Uju Okoli   16     SS3

# 20. What is subquery?
A subquery, or nested query, is a query placed within another SQL query. There are many different scenarios where you may want to include a query in the WHERE, FROM, or SELECT clauses of your main query. Subqueries is an essential tool in SQL, as it allows manipulating the data in very advanced ways without the need of any external scripts, and especially important when your tables stractured in such a way that you can't be joined directly.

In our example, I'm trying to join between a table that holds cities basic details (id, name, population, capital, facts_id), to a table that holds more attributes (facts). The problem is that while the first table holds one row for each country, the key in the second table is id+fact_id, so if we do a regular join, the result would be a Cartesian product, and each country's basic details would appear as many times as the country appears in the attributes table. The problem with of course is that the average would be skewed towards country's that appear many times in the attribute table.

The solution, is to use a subquery. We would need to group the attributes table, to a different key. 

In [40]:
data_path3 = 'lastdatabase.db'
connec = sqlite3.connect(data_path3)

def execute_sql(connec, query):
    cursor = connec.cursor()
    try:
        cursor.execute(query)
        connec.commit()
        print("SQL Executed Successfully")
    except Error as e:
        print (f"The error '{e}' occurred")

In [41]:
gallery_table = """
CREATE TABLE galleries(
    id integer PRIMARY KEY, 
    city text 
    );

"""
paintings_table = """
CREATE TABLE paintings(
    id integer PRIMARY KEY, 
    name text, 
    gallery_id integer, 
    price integer 
    );

"""

agents_table = """
CREATE TABLE sales_agents(
    id integer PRIMARY KEY, 
    lName text, 
    fName text, 
    gallery_id integer, 
    agency_fee integer
    );

"""

managers_table = """
CREATE TABLE managers(
    id integer PRIMARY KEY, 
    gallery_id integer
    );

"""

In [42]:
## Commit create tables
execute_sql(connec, gallery_table)
execute_sql(connec, paintings_table)
execute_sql(connec, agents_table)
execute_sql(connec, managers_table)

SQL Executed Successfully
SQL Executed Successfully
SQL Executed Successfully
SQL Executed Successfully


In [43]:
def insertIntoGallery(id,city):
    try:
        connec = sqlite3.connect('lastdatabase.db')
        cursor = connec.cursor()
        print("Connected to SQLite")

        sqlite_insert = """INSERT INTO galleries(id,city) 
                          VALUES (?,?);"""

        data_tuple = (id,city)
        cursor.execute(sqlite_insert, data_tuple)
        connec.commit()
        print("Python Variables inserted successfully into gallery table")

        cursor.close()

    except sqlite3.Error as error:
         print("Failed to insert Python variable into sqlite table", error)
    finally:
        if connec:
            connec.close()
            print("The SQLite connection is closed")


insertIntoGallery(1, 'Nigeria')
insertIntoGallery(2, 'Kenya')
insertIntoGallery(3, 'Ghana')



Connected to SQLite
Python Variables inserted successfully into gallery table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into gallery table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into gallery table
The SQLite connection is closed


In [44]:
def insertIntoPaint(id,name,gallery_id,price):
    try:
        connec = sqlite3.connect('lastdatabase.db')
        cursor = connec.cursor()
        print("Connected to SQLite")

        sqlite_insert = """INSERT INTO paintings(id,name,gallery_id,price) 
                          VALUES (?,?,?,?);"""

        data_tuple = (id,name,gallery_id,price)
        cursor.execute(sqlite_insert, data_tuple)
        connec.commit()
        print("Python Variables inserted successfully into paintings table")

        cursor.close()

    except sqlite3.Error as error:
        print("Failed to insert Python variable into sqlite table", error)
    finally:
        if connec:
            connec.close()
            print("The SQLite connection is closed")


insertIntoPaint(1, 'Words On Walls', 3, 100000)
insertIntoPaint(2, 'Oceans view', 1, 5000000)
insertIntoPaint(3, 'Hiding Galaxies', 1, 750000)
insertIntoPaint(4, 'Far From the Madding Crowd', 2, 2000000)
insertIntoPaint(5, 'Accelerated Waves', 2, 350000)

Connected to SQLite
Python Variables inserted successfully into paintings table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into paintings table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into paintings table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into paintings table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into paintings table
The SQLite connection is closed


In [45]:
def insertIntoAgent(id,lname,fname,gallery_id,agency_fee):
    try:
        connec = sqlite3.connect('lastdatabase.db')
        cursor = connec.cursor()
        print("Connected to SQLite")

        sqlite_insert = """INSERT INTO sales_agents(id,lname,fname,gallery_id,agency_fee)  
                          VALUES (?,?,?,?,?);"""

        data_tuple = (id,lname,fname,gallery_id,agency_fee)
        cursor.execute(sqlite_insert, data_tuple)
        connec.commit()
        print("Python Variables inserted successfully into sales_agents table")

        cursor.close()

    except sqlite3.Error as error:
        print("Failed to insert Python variable into sqlite table", error)
    finally:
        if connec:
            connec.close()
            print("The SQLite connection is closed")


insertIntoAgent(1, 'Chinenye','James', 2, 5000)
insertIntoAgent(2, 'Winner','Emeto', 3, 40000)
insertIntoAgent(3, 'Hilda', 'Onuoha', 2, 20000)
insertIntoAgent(4, 'Betty', 'Sharp', 1, 100000)
insertIntoAgent(5, 'Doubting', 'Thomas', 3, 12000)

Connected to SQLite
Python Variables inserted successfully into sales_agents table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into sales_agents table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into sales_agents table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into sales_agents table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into sales_agents table
The SQLite connection is closed


In [46]:

def insertIntoManagers(id,gallery_id):
    try:
        connec = sqlite3.connect('lastdatabase.db')
        cursor = connec.cursor()
        print("Connected to SQLite")

        sqlite_insert = """INSERT INTO managers(id,gallery_id) 
                          VALUES (?,?);"""

        data_tuple = (id,gallery_id) 
        cursor.execute(sqlite_insert, data_tuple)
        connec.commit()
        print("Python Variables inserted successfully into managers table")

        cursor.close()

    except sqlite3.Error as error:
        print("Failed to insert Python variable into sqlite table", error)
    finally:
        if connec:
            connec.close()
            print("The SQLite connection is closed")


insertIntoManagers(1, 2)
insertIntoManagers(2, 3)
insertIntoManagers(4, 1)

Connected to SQLite
Python Variables inserted successfully into managers table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into managers table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into managers table
The SQLite connection is closed


In [47]:
view_gallery = pd.read_sql("""SELECT *
                        FROM galleries;""", connec)
view_gallery

id     city
0   1  Nigeria
1   2    Kenya
2   3    Ghana

In [48]:
view_paintings = pd.read_sql("""SELECT *
                        FROM paintings;""", connec)
view_paintings


id                        name  gallery_id    price
0   1              Words On Walls           3   100000
1   2                 Oceans view           1  5000000
2   3             Hiding Galaxies           1   750000
3   4  Far From the Madding Crowd           2  2000000
4   5           Accelerated Waves           2   350000

In [49]:


view_agents = pd.read_sql("""SELECT *
                        FROM sales_agents;""", connec)
view_agents

id     lName   fName  gallery_id  agency_fee
0   1  Chinenye   James           2        5000
1   2    Winner   Emeto           3       40000
2   3     Hilda  Onuoha           2       20000
3   4     Betty   Sharp           1      100000
4   5  Doubting  Thomas           3       12000

In [50]:
view_managers = pd.read_sql("""SELECT *
                        FROM managers;""", connec)
view_managers

id  gallery_id
0   1           2
1   2           3
2   4           1

## One of the most straightforward subquery use cases is to include it in the WHERE clause to filter results. For example, if you wanted to see information about only those sales agents who received a higher-than-average agency fee last month, you could use the following SQL query:

In [51]:
sub_query = """
SELECT *
FROM sales_agents
WHERE agency_fee > 
(SELECT AVG(agency_fee)
FROM sales_agents);

"""
view_suquery = pd.read_sql(sub_query, connec)
view_suquery

id   lName  fName  gallery_id  agency_fee
0   2  Winner  Emeto           3       40000
1   4   Betty  Sharp           1      100000

# 21. What are the types of subquery?
## A. Scalar Subqueries
When a subquery returns a single value, or exactly one row and exactly one column, we call it a scalar subquery. This type of subquery is frequently used in the WHERE clause to filter the results of the main query. The subquery in our previous example is a scalar subquery, as it returns a single value (i.e. the average agency fee).

Scalar subqueries can also be used in the main query’s SELECT statement. For example, let’s say we want to see the average price of all our paintings next to the price of each painting.

In [52]:
scalar_type = """
SELECT name AS painting,price,
(SELECT AVG(price)
FROM paintings) 
AS avg_price
FROM paintings;

"""
view_scalar = pd.read_sql(scalar_type, connec)
view_scalar

painting    price  avg_price
0              Words On Walls   100000  1640000.0
1                 Oceans view  5000000  1640000.0
2             Hiding Galaxies   750000  1640000.0
3  Far From the Madding Crowd  2000000  1640000.0
4           Accelerated Waves   350000  1640000.0

## B. Multiple-Row Subqueries
If your subquery returns more than one row, it can be referred to as a multiple-row subquery. Note that this subquery type includes (1) subqueries that return one column with multiple rows (i.e. a list of values) and (2) subqueries that return multiple columns with multiple rows (i.e. tables).

In [53]:
multiple_type = """
SELECT AVG(agency_fee)
FROM sales_agents
WHERE id NOT IN (SELECT id FROM managers);

"""
view_multiple = pd.read_sql(multiple_type, connec)
view_multiple

AVG(agency_fee)
0          16000.0

## C. Correlated Subqueries
There are also SQL subqueries where the inner query relies on information obtained from the outer query. These are correlated subqueries. Because of the interdependence between the main query and the inner query, this type of subquery can be more challenging to understand. 

In [54]:
correlated_type = """
SELECT city, 
(SELECT count(*)
FROM paintings p
WHERE g.id = p.gallery_id) total_paintings
FROM galleries g;

"""
view_correlated = pd.read_sql(correlated_type, connec)
view_correlated

city  total_paintings
0  Nigeria                2
1    Kenya                2
2    Ghana                1

### Finally, correlated subqueries can also be used in the WHERE statement. For instance, let’s say we want to get information about those sales agents whose agency fee was equal or higher than the average fee for their gallery. We can run the following query to get the required result:

In [55]:
still_correlated = """
SELECT lname,fname,agency_fee
FROM sales_agents sa1
WHERE sa1.agency_fee >= (SELECT avg(agency_fee)
                         FROM sales_agents sa2 
                         WHERE sa2.gallery_id = sa1.gallery_id);

"""
view_still = pd.read_sql(still_correlated, connec)
view_still

lName   fName  agency_fee
0  Winner   Emeto       40000
1   Hilda  Onuoha       20000
2   Betty   Sharp      100000

# 22. What is a stored procedure?
A stored procedure is a prepared SQL code that you can save, so the code can be reused over and over again. So if you have an SQL query that you write over and over again, save it as a stored procedure, and then just call it to execute it. You can also pass parameters to a stored procedure, so that the stored procedure can act based on the parameter value(s) that is passed.

In [56]:
my_procedure = """
CREATE PROCEDURE SelectAllFacts
AS
SELECT * FROM facts
GO;

"""

execute_procedure = """
EXEC 
    SelectAllFacts;
"""



# 23. What is a trigger?
 A trigger is a stored procedure in database which automatically invokes whenever a special event in the database occurs. For example, a trigger can be invoked when a row is inserted into a specified table or when certain table columns are being updated.

In [57]:
data_path2 = 'womentechsters.db'
conn = sqlite3.connect(data_path2)
def execute_sql(conn, query):
    cursor = conn.cursor()
    try:
        cursor.execute(query)
        conn.commit()
        print("SQL Executed Successfully")
    except Error as e:
        print (f"The error '{e}' occurred")

score_table = """
CREATE TABLE Scores(
    id integer PRIMARY KEY, 
    name text,
    subj1 integer, 
    subj2 integer,
    subj3 integer,
    total integer,
    per integer
    );

"""

execute_sql(conn, score_table)

SQL Executed Successfully


In [58]:
trigger_it = """
CREATE trigger stud_marks 
    BEFORE INSERT 
    ON Scores 
    FOR EACH ROW set Scores.total = Scores.subj1 + Scores.subj2 + Scores.subj3, Scores.per = Scores.total * 60 / 100

"""


In [59]:
execute_sql(conn, trigger_it)

The error 'near "set": syntax error' occurred


In [60]:
def insertIntoScores(id,name,subj1,subj2,subj3,total,per):
    try:
        conn = sqlite3.connect('womentechsters.db')
        cursor = conn.cursor()
        print("Connected to SQLite")

        sqlite_insert = """INSERT INTO Scores(id,name,subj1,subj2,subj3,total,per)
                          VALUES (?,?,?,?,?,?,?);"""

        data_tuple = (id,name,subj1,subj2,subj3,total,per)
        cursor.execute(sqlite_insert, data_tuple)
        conn.commit()
        print("Python Variables inserted successfully into scores table")

        cursor.close()

    except sqlite3.Error as error:
        print("Failed to insert Python variable into sqlite table", error)
    finally:
        if conn:
            conn.close()
            print("The SQLite connection is closed")


insertIntoScores(1, "Alimi Badmus", 80, 68, 40, 0, 0)



Connected to SQLite
Python Variables inserted successfully into scores table
The SQLite connection is closed


In [61]:
view_scores = pd.read_sql("""SELECT *
                        FROM Scores;""", conn)
view_scores

id          name  subj1  subj2  subj3  total  per
0   1  Alimi Badmus     80     68     40      0    0

# 24. What is the difference between DELETE and TRUNCATE command?
Delete and truncate both commands can be used to delete data of the table. 

Delete is a Data Manipulation Language (DML) command whereas truncate is Data Definition Language (DDL) command. Truncate can be used to delete the entire data of the table without maintaining the integrity of the table. On the other hand , delete statement can be used for deleting the specific data. 

In [62]:
## TRUNCATE Query
trunc = """
TRUNCATE TABLE Scores;
"""

## Delete
del_table = """
DELETE FROM paintings
WHERE name="Words On Walls";

"""

In [63]:
view_paintings = pd.read_sql("""SELECT *
                        FROM paintings;""", connec)
view_paintings

id                        name  gallery_id    price
0   1              Words On Walls           3   100000
1   2                 Oceans view           1  5000000
2   3             Hiding Galaxies           1   750000
3   4  Far From the Madding Crowd           2  2000000
4   5           Accelerated Waves           2   350000

# 25. What is a constraint?
SQL constraints are used to specify rules for data in a table. Constraints can be specified when the table is created with the CREATE TABLE statement, or after the table is created with the ALTER TABLE statement.
Examples of constraints include:
- NOT NULL
- PRIMARY KEY
- UNIQUE
- FOREIGN KEY

Some of the tables above were created with constraints

# 26. What is data integrity?
Data Integrity is a process or a set of practices that ensures the security, accuracy, and overall quality of data. It is a data that has no redundance

# 27. What is auto increment?
Auto-increment allows a unique number to be generated automatically when a new record is inserted into a table. Often this is the primary key field that we would like to be created automatically every time a new record is inserted.

In [64]:
## Example of auto increment
scoretable = """
CREATE TABLE Score(id int AUTO_INCREMENT,name text,age integer,PRIMARY KEY (id));

"""

execute_sql(conn, scoretable)

def insertIntoScore(name,age):
    try:
        conn = sqlite3.connect('womentechsters.db')
        cursor = conn.cursor()
        print("Connected to SQLite")

        sqlite_insert = """INSERT INTO Score(name,age)
                          VALUES (?,?);"""

        data_tuple = (name,age)
        cursor.execute(sqlite_insert, data_tuple)
        conn.commit()
        print("Python Variables inserted successfully into score table")

        cursor.close()

    except sqlite3.Error as error:
        print("Failed to insert Python variable into sqlite table", error)
    finally:
        if conn:
            conn.close()
            print("The SQLite connection is closed")


insertIntoScore("Alimi Badmus", 34)
insertIntoScore("Daney Moss", 35)
insertIntoScore("Creek Beans", 45)
insertIntoScore("Nuella Chick", 53)





SQL Executed Successfully
Connected to SQLite
Python Variables inserted successfully into score table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into score table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into score table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into score table
The SQLite connection is closed


In [65]:
view_me = pd.read_sql("""SELECT *
                        FROM Score;""", conn)
view_me

id          name  age
0  None  Alimi Badmus   34
1  None    Daney Moss   35
2  None   Creek Beans   45
3  None  Nuella Chick   53

# 28. What are local and global variables and their difference?
## Local variable: Local variable is declared inside a function. Local variables are created when the function has started execution and is lost when the function terminates. 
- A user declares the local variable.
- By default, a local variable starts with @.
- Every local variable scope has the restriction to the current batch or procedure within any given session.
## Global variable: Global variable is declared outside the function. Global variable is created as execution starts and is lost when the program ends.
- The system maintains the global variable. A user cannot declare them.
- The global variable starts with @@
- It stores session related information.

In [ ]:
## Example of global declaration

see_lang = """
SELECT @@language
"""
run_it = pd.read_sql(see_lang,connection)

## Examples of local declaration
local_declare = """
DECLARE @TestVariable AS VARCHAR(100)='Save Our Planet'
PRINT @TestVariable
"""
runs_it = pd.read_sql(local_declare,connection)

# 29. What is the difference between cluster and non-cluster index?
## Clustered Index: A clustered index defines the order in which data is stored in the table which can be sorted in only one way. So, there can be an only a single clustered index for every table.
## Unclustered Index: A Non-clustered index stores the data at one location and indices at another location. The index contains pointers to the location of that data. A single table can have many non-clustered indexes as an index in the non-clustered index is stored in different places.

In [66]:
products = """
CREATE TABLE orders(
    order_id integer PRIMARY KEY, 
    sales_id int auto_increment, 
    order_qty integer, 
    product_id integer,
    unit_price integer
    );

"""
execute_sql(connec, products)

SQL Executed Successfully


In [67]:
def insertIntoOrders(sales_id,order_qty,product_id,unit_price):
    try:
        connec = sqlite3.connect('lastdatabase.db')
        cursor = connec.cursor()
        print("Connected to SQLite")

        sqlite_insert = """INSERT INTO orders(sales_id,order_qty,product_id,unit_price) 
                          VALUES (?,?,?,?);"""

        data_tuple = (sales_id,order_qty,product_id,unit_price) 
        cursor.execute(sqlite_insert, data_tuple)
        connec.commit()
        print("Python Variables inserted successfully into orders table")

        cursor.close()

    except sqlite3.Error as error:
        print("Failed to insert Python variable into sqlite table", error)
    finally:
        if connec:
            connec.close()
            print("The SQLite connection is closed")


insertIntoOrders(211,2,611,100)
insertIntoOrders(211,2,613,50)
insertIntoOrders(211,2,614,5)
insertIntoOrders(211,1,612,10)
insertIntoOrders(211,1,689,120)
insertIntoOrders(211,3,615,1000)
insertIntoOrders(213,3,617,55)
insertIntoOrders(213,3,620,400)
insertIntoOrders(215,1,601,7500)

Connected to SQLite
Python Variables inserted successfully into orders table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into orders table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into orders table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into orders table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into orders table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into orders table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into orders table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into orders table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into orders table
The SQLite connection is closed


In [68]:
view_orders = pd.read_sql("""SELECT *
                        FROM orders;""", connec)
view_orders

order_id  sales_id  order_qty  product_id  unit_price
0         1       211          2         611         100
1         2       211          2         613          50
2         3       211          2         614           5
3         4       211          1         612          10
4         5       211          1         689         120
5         6       211          3         615        1000
6         7       213          3         617          55
7         8       213          3         620         400
8         9       215          1         601        7500

In [69]:
## Clustered Index
clustered = """
SELECT product_id, unit_price
FROM orders
WHERE sales_id = 6;
"""

## Non-clustered Index
non_clustered = """
CREATE INDEX myIndex 
ON orders(product_id, order_qty);
"""

non_clust = """
SELECT product_id, order_qty
FROM orders
WHERE product_id = 617;
"""


In [70]:
view_nonclus = pd.read_sql(non_clust, connec)
view_nonclus

product_id  order_qty
0         617          3

In [71]:
view_clus = pd.read_sql(clustered, connec)
view_clus

Empty DataFrame
Columns: [product_id, unit_price]
Index: []

# 30. What is a Datawarehouse?
A data warehouse is a large collection of business data used to help an organization make decisions.

# 33. What is user defined function?


In [72]:
## Creating new table
testing = """
 CREATE TABLE TEST(
    num1 INT,
    num2 INT
    );

"""
execute_sql(connec, testing)

SQL Executed Successfully


In [73]:
def insertIntoTest(num1,num2):
    try:
        connec = sqlite3.connect('lastdatabase.db')
        cursor = connec.cursor()
        print("Connected to SQLite")

        sqlite_insert = """INSERT INTO TEST(num1,num2) 
                          VALUES (?,?);"""

        data_tuple = (num1,num2) 
        cursor.execute(sqlite_insert, data_tuple)
        connec.commit()
        print("Python Variables inserted successfully into TEST table")

        cursor.close()

    except sqlite3.Error as error:
        print("Failed to insert Python variable into sqlite table", error)
    finally:
        if connec:
            connec.close()
            print("The SQLite connection is closed")


insertIntoTest(1,2)
insertIntoTest(2,3)
insertIntoTest(4,5)

Connected to SQLite
Python Variables inserted successfully into TEST table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into TEST table
The SQLite connection is closed
Connected to SQLite
Python Variables inserted successfully into TEST table
The SQLite connection is closed


## We will create a function that returns a new table with an extra column. This extra column will contain the sum of numbers in the column num1 and column num2.

In [74]:
## Table-Valued Functions
the_func = """
CREATE FUNCTION table_valued_func()
    RETURNS TEST
    AS
    RETURN
-- calculate sum
SELECT num1,num2,num1+num2 
AS 'SUM'
FROM TEST;

"""


def upper(string):
    return str(string).upper()

connec = sqlite3.connect('lastdatabase.db')
cursor = connec.cursor()
connec.create_function("upper", 1, upper)
cursor.execute(the_func)
#print cursor.fetchone()




OperationalError: near "FUNCTION": syntax error

In [75]:
connec.create_function(the_func,1,connec)

In [76]:
execute_sql(connec, the_func)

The error 'near "FUNCTION": syntax error' occurred


# 34. What are the types of user defined functions?
There are two main types of user-defined functions in SQL based on the data they return:

- Scalar functions: These types of functions return a single value, i.e float, int, varchar, datetime, etc.

Table-Valued functions: These functions return tables.

In [77]:
## Scalar Functions
the_func2 = """

CREATE FUNCTION scalar_func(
    @a AS INT, -- parameter a
    @b AS INT -- parameter b
)
RETURNS INT 
AS
BEGIN
RETURN @a + @b 
END;
"""
execute_sql(connec, the_func2)

view_func2 = pd.read_sql(the_func2, connec)
view_func2

The error 'near "FUNCTION": syntax error' occurred


DatabaseError: Execution failed on sql '

CREATE FUNCTION scalar_func(
    @a AS INT, -- parameter a
    @b AS INT -- parameter b
)
RETURNS INT 
AS
BEGIN
RETURN @a + @b 
END;
': near "FUNCTION": syntax error

# 35. What is a collation?
Collation is a set of rules that tell database engine how to compare and sort the character data in SQL Server. Collation can be set at different levels in SQL Server. Below are the three levels:

- SQL Server level
- Database level
- Column level

# 36. What are the different types of collation sensitivity?
- Case sensitivity – A and a, B and b, etc.
- Accent sensitivity – a and á, o and ó, etc.
- Kana Sensitivity – When Japanese kana characters Hiragana and Katakana are treated differently, it is called Kana sensitive.
- Width sensitivity – A single-byte character (half-width) and the same character represented as a double-byte character (full-width) are treated differently than it is width sensitive. 

# 37. What are the advantages and disadvantages of stored procedure
## Stored Procedure Advantages
Followings are the advantages of using MySQL Stored Procedures:−

1. Increasing the performance of applications − As we know that after creating the stored procedure it is compiled and stored in the database. But MySQL implements stored procedures slightly different which helps in increasing the performance of the applications. MySQL stored procedures are compiled on demand. 

2. Fast − MySQL Stored procedures are fast because the MySQL server takes some advantage of caching. Another reason for its speed is that it makes the reduction in network traffic. Suppose, if we have a repetitive task that requires checking, looping, multiple statements, and no user interaction, does it with a single call to a procedure that's stored on the server.

3. Portable − MySQL Stored procedures are portable because when we write our stored procedure in SQL, we know that it will run on every platform that MySQL runs on, without obliging us to install an additional runtime-environment package or set permissions for program execution in the operating system.

4. Reusable and transparent − Stored procedures expose the database interface to all applications so that developers don’t have to develop functions that are already supported in stored procedures. Hence, we can say that MySQL stored procedures are reusable and transparent.

5. Secure − MySQL stored procedures are secure because the database administrator can grant appropriate permissions to applications that access stored procedures in the database without giving any permissions on the underlying database tables.

## Stored Procedure Disadvantages
Followings are the advantages of using MySQL Stored Procedures −

1. Memory usage increased − If we use many stored procedures, the memory usage of every connection that is using those stored procedures will increase substantially.

2. Restricted for complex business logic − Actually, stored procedure’s constructs are not designed for developing complex and flexible business logic.

3. Difficult to debug − It is difficult to debug stored procedures. Only a few database management systems allow you to debug stored procedures. Unfortunately, MySQL does not provide facilities for debugging stored procedures.

4. Difficult to maintain − It is not easy to develop and maintain stored procedures. Developing and maintaining stored procedures are often required a specialized skill set that not all application developers possess. This may lead to problems in both application development and maintenance phases.


# 38. What is online transfer processing (OLTP)?
OLTP (online transaction processing) is a class of software programs capable of supporting transaction-oriented applications on the Internet. In computing, a transaction is a sequence of discrete information exchanges that are treated as a unit. Many everyday acts involve OLTP, including online banking, online shopping and even in-store shopping when the point of sale (POS) terminal is tied to inventory management software.

Two important characteristics of an OLTP system are 
- concurrency, and 
- atomicity.

# 39. What is a CLAUSE?
Clauses are in-built functions available to us in SQL. With the help of clauses, we can deal with data easily stored in the table.

Clauses help us filter and analyze data quickly. When we have large amounts of data stored in the database, we use Clauses to query and get data required by the user.

Some of the examples of clauses are – where, and, or, like, top, etc.

In [78]:
## WHERE Clause
non_clust = """
SELECT product_id, order_qty
FROM orders
WHERE product_id = 617;
"""
view_nonclus = pd.read_sql(non_clust, connec)
view_nonclus

product_id  order_qty
0         617          3

# 40. What is recursive stored procedure?
Recursion is when a procedure calls itself until a stop condition is reached. However, due to limit on nesting levels, recursion is limited to 32 calls. The classic example of recursion is calculating a factorial of a number. 

# 41. What is Union, minus and Interact commands?
## Union:
The UNION operator is used to combining the results of two tables, and it eliminates duplicate rows from the tables.
## Minus:
The MINUS operator is used to returning rows from the first query but not from the second query. The matching records of the first and second query and other rows from the first query will be displayed as a result set.
## Intersect:
The INTERSECT operator is used to return rows returned by both the queries.

In [79]:
insertIntoPaint(6, 'Ascending', 4, 2580000)


Connected to SQLite
Python Variables inserted successfully into paintings table
The SQLite connection is closed


In [80]:
## Union
union_query = """
SELECT * FROM managers
UNION
SELECT * FROM galleries;  

"""

view_minus = pd.read_sql(union_query, connec)
view_minus

id gallery_id
0   1          2
1   1    Nigeria
2   2          3
3   2      Kenya
4   3      Ghana
5   4          1

In [81]:
## Minus
minus_query = """
SELECT * FROM managers
MINUS
SELECT * FROM galleries;

"""
view_minus = pd.read_sql(minus_query, connec)
view_minus

DatabaseError: Execution failed on sql '
SELECT * FROM managers
MINUS
SELECT * FROM galleries;

': near "SELECT": syntax error

In [82]:
## Intersect
intersect_query = """
SELECT * FROM managers
INTERSECT
SELECT * FROM galleries;

"""
view_intersect = pd.read_sql(intersect_query, connec)
view_intersect

Empty DataFrame
Columns: [id, gallery_id]
Index: []

# 42A. What is an ALIAS Command?
SQL aliases are used to give a table, or a column in a table, a temporary name.

- Aliases are often used to make column names more readable.
- An alias only exists for the duration of that query.
- An alias is created with the AS keyword.

A typical example is on 21 above

# 42B.  ALIAS name can be given to a table or column. This alias name can be referred in WHERE clause to identify the table or column, Show


In [83]:
scalar_type = """
SELECT name AS painting,price,
(SELECT AVG(price)
FROM paintings) 
AS avg_price
FROM paintings;

"""
view_scalar = pd.read_sql(scalar_type, connec)
view_scalar

painting    price     avg_price
0              Words On Walls   100000  1.796667e+06
1                 Oceans view  5000000  1.796667e+06
2             Hiding Galaxies   750000  1.796667e+06
3  Far From the Madding Crowd  2000000  1.796667e+06
4           Accelerated Waves   350000  1.796667e+06
5                   Ascending  2580000  1.796667e+06

# 43. What is the difference between TRUNCATE and DROP statements?
TRUNCATE removes all the rows from the table, and it cannot be rolled back. 
DROP command removes a table from the database and operation cannot be rolled back.

In [84]:
drop_table = """
DROP Fellows;

"""
execute_sql(conn, drop_table)

confirm = pd.read_sql(drop_table, conn)
confirm

The error 'near "Fellows": syntax error' occurred


DatabaseError: Execution failed on sql '
DROP Fellows;

': near "Fellows": syntax error

# 44. What are aggregate and scalar functions?
## Aggregate functions:
These functions are used to do operations from the values of the column and a single value is returned.
- AVG()
- COUNT()
- FIRST()
- LAST()
- MAX()
- MIN()
- SUM()
## Scalar functions:
These functions are based on user input, these too returns single value.
- UCASE()
- LCASE()
- MID()
- LEN()
- ROUND()
- NOW()
- FORMAT()

# 45. How can you create an empty table from an existing table?

In [85]:
newtable = """
CREATE TABLE factsb AS
(SELECT * FROM facts WHERE (id > 100));

"""
execute_sql(connection, newtable)

The error 'near "(": syntax error' occurred


In [86]:
answer = pd.read_sql(newtable, connection)
answer

DatabaseError: Execution failed on sql '
CREATE TABLE factsb AS
(SELECT * FROM facts WHERE (id > 100));

': near "(": syntax error

# 46. How to fetch common records from two tables?

In [87]:
no_46 = """
Select id from studentrack 
INTERSECT Select id from track
"""
execute_sql(conn, no_46)

SQL Executed Successfully


In [88]:
out_46 = pd.read_sql(conn, no_46)
out_46

ArgumentError: Could not parse rfc1738 URL from string '
Select id from studentrack 
INTERSECT Select id from track
'

# 47A. How to fetch alternate records from a table?

# 47B. How do you display even numbers-.

In [89]:
# To display even numbers-.

even_num = """
Select id,area 
from (Select id, area from facts) 
where mod(id,2)=0

"""
execute_sql(connection, even_num)


SQL Executed Successfully


In [90]:
answ = pd.read_sql(even_num, connection)
answ

id     area
0      2  28748.0
1      4    468.0
2      6    442.0
3      8  29743.0
4     10  83871.0
..   ...      ...
125  252      7.0
126  254   5860.0
127  256      NaN
128  258      NaN
129  260      NaN

[130 rows x 2 columns]

# 47C. How do you display odd numbers-.

In [91]:
# To display odd numbers-.
odd_num = """
Select id,area 
from (Select id,area from facts) 
where mod(id,2)=1

"""
execute_sql(connection, odd_num)

SQL Executed Successfully


In [92]:
answr = pd.read_sql(odd_num, connection)
answr

id       area
0      1   652230.0
1      3  2381741.0
2      5  1246700.0
3      7  2780400.0
4      9  7741220.0
..   ...        ...
126  253        5.0
127  255   266000.0
128  257        NaN
129  259        NaN
130  261        NaN

[131 rows x 2 columns]

# 48. How to select unique records from a table?

In [93]:
dist = """
Select DISTINCT capital, facts_id, name from cities;
"""
execute_sql(connection, dist)

SQL Executed Successfully


In [94]:
anw = pd.read_sql(dist, connection)
anw

capital facts_id          name
0          1      216    Oranjestad
1          1        6  Saint John'S
2          1      184     Abu Dhabi
3          0      184         Dubai
4          0      184       Sharjah
..       ...      ...           ...
392        1      167       Mbabane
393        1      193         Sanaa
394        0      193          Aden
395        1      194        Lusaka
396        1      195        Harare

[397 rows x 3 columns]

# 49. What is the command used to fetchfirst 5 characters of the string?

In [95]:
fetch_5 = """
Select SUBSTRING(name,1,5) 
as studentname 
from Student;
"""

view_fetch = pd.read_sql(fetch_5, conn)
view_fetch

studentname
0       Amara
1       Chine
2       Rukay
3       Chiom
4       Uju O

# 50. Which operator is used in query forpattern matching?
LIKE operator is used for pattern matching, and it can be used as -.

- % - Matches zero or more characters.
- _(Underscore) – Matching exactly one character.
Example:

In [96]:
match_1 = """
Select * from Student 
where name 
like 'c%'
"""

match_2 = """
Select * from Student
where name 
like 'hol_'
"""

view_match = pd.read_sql(match_1, conn)
view_match


id            name  age class_s
0   2  Chinenye James   11     SS2
1   4      Chioma Obi   12     SS1

In [97]:
view_match2 = pd.read_sql(match_2, conn)
view_match2

Empty DataFrame
Columns: [id, name, age, class_s]
Index: []